packages/liveblocks-node/src/thread-notification-helpers.ts

```diff
@@ -0,0 +1,318 @@
+  };
+
+export type ResolvedCommentAuthorData = {
+  id: string;
+  name: string;
+  avatar?: string;
+};
+
+/** @internal */
+const resolveAuthorsInComments = async <U extends BaseUserMeta>({
+  comments,
+  resolveUsers,
+}: {
+  comments: CommentDataWithBody[];
+  resolveUsers?: (
+    args: ResolveUsersArgs
+  ) => OptionalPromise<(U["info"] | undefined)[] | undefined>;
+}): Promise<Map<string, U["info"]>> => {
+  const resolvedAuthors = new Map<string, U["info"]>();
+  if (!resolveUsers) {
+    return resolvedAuthors;
+  }
+
+  const userIds = comments.map((c) => c.userId);
+  const users = await resolveUsers({ userIds });
+
+  for (const [index, userId] of userIds.entries()) {
+    const user = users?.[index];
+    if (user) {
+      resolvedAuthors.set(userId, user);
+    }
+  }
+
+  return resolvedAuthors;
+};
+
+export type ResolvedCommentData = {
+  id: string;
+  threadId: string;
+  roomId: string;
+  author: ResolvedCommentAuthorData;
+  createdAt: Date;
+  commentUrl?: string;
+} & ({ body: string } | { body: CommentBodyJson });
+
+export type ResolvedUnreadRepliesData = {
+  type: "unreadReplies";
+  comments: ResolvedCommentData[];
+};
+export type ResolvedUnreadMentionData = {
+  type: "unreadMention";
+  comments: ResolvedCommentData[];
+};
+export type ThreadNotificationResolvedData = (
+  | ResolvedUnreadRepliesData
+  | ResolvedUnreadMentionData
+) & {
+  roomInfo: DRI;
+};
+
+/**
+ *
+ * Get thread notification resolved data helper.
+ *
+ * It returns either an object containing a list of unread replies for a thread
+ * or either an object containing a list of the last unread comment where the notification
+ * receiver was mentioned in.
+ *
+ * @param params.client Liveblocks node client
+ * @param params.event The thread notification event
+ * @returns A thread notification resolved data object
+ *
+ * @example Unread replies:
+ * {
+ *  type: "unreadReplies",
+ *  comments: [unread_comment_0, unread_comment_1, ...],
+ *  roomName: "acme"
+ * }
+ *
+ * @example Unread mention:
+ * {
+ *  type: "unreadMention",
+ *  comments: [unread_comment_with_mention],
+ *  roomName: "acme"
+ * }
+ */
+export async function getThreadNotificationResolvedData(params: {
+  client: Liveblocks;
+  event: ThreadNotificationEvent;
+  options?: GetThreadNotificationResolvedOptions<BaseUserMeta>;
+}): Promise<ThreadNotificationResolvedData> {
+  const { client, event, options } = params;
+  const { roomId } = event.data;
+
+  const roomInfos = options?.resolveRoomInfo
+    ? await options.resolveRoomInfo({ roomId })
+    : undefined;
+
+  const { type, comments } = await getThreadNotificationData({
+    client,
+    event,
+  });
+
+  const resolvedAuthors = await resolveAuthorsInComments({
+    comments,
+    resolveUsers: options?.resolveUsers,
+  });
+
+  const unreadComments = await Promise.all(
+    comments.map(async (comment): Promise<ResolvedCommentData> => {
+      const body = await transformCommentBody(comment.body, {
```

Comment:
This leads to `n` calls made to our customers' database (or auth provider) to retrieve the users information. It will causes performance issues and invocation timeouts on customer webhook handlers.

`resolveUsers` should be called a single time during the invocation of this function.

Reply:
Agreed! thanks for catching this 👍🏻